File: scripts/previous_in_vitro/matlab/synaptic_input_output_analysis.m
% synaptic_analysis.m

% Synaptic Input-Output Analysis in MATLAB

%% Function to Load Data
function data = load_data(file_path)
    % Load synaptic input-output data from a CSV file.
    % Args:
    %     file_path (str): Path to the input data file.
    % Returns:
    %     data (table): Loaded data as a MATLAB table.
    % Raises:
    %     Error if file does not exist or data format is incorrect.
    
    % Check if the file exists
    if ~isfile(file_path)
        error('FileNotFoundError: The file "%s" does not exist.', file_path);
    end
    
    % Try loading the file
    try
        data = readtable(file_path);
        if ~all(ismember({'Input', 'Output'}, data.Properties.VariableNames))
            error('ValueError: The input data must contain "Input" and "Output" columns.');
        end
    catch ME
        rethrow(ME);
    end
end

%% Function to Perform Synaptic Input-Output Analysis
function [fitresult, gof] = analyze_synaptic_io(data, model)
    % Analyze synaptic input-output relationships using curve fitting.
    % Args:
    %     data (table): Synaptic data containing input and output columns.
    %     model (str): Model type for curve fitting ('sigmoid' or 'linear').
    % Returns:
    %     fitresult (cfit): Fitted curve object.
    %     gof (struct): Goodness of fit structure.
    
    % Validate the model parameter
    if ~ismember(model, {'sigmoid', 'linear'})
        error('ValueError: Model must be "sigmoid" or "linear".');
    end
    
    % Extract input and output data
    input_data = data.Input;
    output_data = data.Output;
    
    % Define model functions for curve fitting
    sigmoid = @(a, b, c, x) c ./ (1 + exp(-(x - a) / b));
    linear = @(m, c, x) m * x + c;
    
    % Fit the selected model to the data
    try
        switch model
            case 'sigmoid'
                [fitresult, gof] = fit(input_data, output_data, ...
                    fittype(sigmoid), 'StartPoint', [0, 1, max(output_data)]);
            case 'linear'
                [fitresult, gof] = fit(input_data, output_data, ...
                    fittype(linear), 'StartPoint', [1, 0]);
        end
    catch ME
        error('FitError: An error occurred during curve fitting: %s', ME.message);
    end
end

%% Function to Plot the Results
function plot_results(data, fitresult, model)
    % Plot synaptic input-output data and the fitted curve.
    % Args:
    %     data (table): Synaptic data containing input and output columns.
    %     fitresult (cfit): Fitted curve object.
    %     model (str): Model type used for curve fitting ('sigmoid' or 'linear').
    
    % Plot data points
    figure;
    scatter(data.Input, data.Output, 'filled');
    hold on;
    
    % Plot the fitted curve
    plot(fitresult, 'r-');
    xlabel('Synaptic Input');
    ylabel('Synaptic Output');
    title(sprintf('Synaptic Input-Output Analysis (%s Model)', model));
    legend('Data', 'Fitted Curve');
    hold off;
end

%% Main Script to Run the Analysis
try
    % Set file path and model type
    file_path = '../../data/synaptic_io_data.csv';  % Ensure the data file exists in the 'data/' directory
    model_type = 'sigmoid';  % Change to 'linear' for a different model
    
    % Load data
    data = load_data(file_path);
    
    % Perform analysis
    [fitresult, gof] = analyze_synaptic_io(data, model_type);
    
    % Display goodness of fit
    disp(gof);
    
    % Plot the results
    plot_results(data, fitresult, model_type);

catch ME
    fprintf('An error occurred: %s\n', ME.message);
end

File: scripts/patch_clamp_whole_cell_in_vitro/python/v1/synaptic_input_output_analysis.py
# synaptic_input_output_analysis.py

# Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import curve_fit
import os

# Function to load data
def load_data(file_path):
    """
    Load synaptic input-output data from a CSV file.
    
    Args:
        file_path (str): Path to the input data file.
    
    Returns:
        data (DataFrame): Loaded data as a Pandas DataFrame.
    Raises:
        FileNotFoundError: If the file is not found at the specified path.
        ValueError: If the data format is incorrect or missing columns.
    """
    if not os.path.exists(file_path):
        raise FileNotFoundError(f"The file '{file_path}' does not exist.")
    
    data = pd.read_csv(file_path)
    if 'Input' not in data.columns or 'Output' not in data.columns:
        raise ValueError("The input data must contain 'Input' and 'Output' columns.")
    
    return data

# Function to perform synaptic input-output analysis
def analyze_synaptic_io(data, model='sigmoid'):
    """
    Analyze synaptic input-output relationships using curve fitting.
    
    Args:
        data (DataFrame): Synaptic data containing input and output columns.
        model (str): The model to use for curve fitting ('sigmoid' or 'linear').
    
    Returns:
        popt (ndarray): Optimal parameters for the fitted curve.
        pcov (ndarray): Covariance of the parameters.
    """
    # Validate model type
    if model not in ['sigmoid', 'linear']:
        raise ValueError("Model must be 'sigmoid' or 'linear'.")
    
    # Define model functions
    def sigmoid(x, a, b, c):
        return c / (1 + np.exp(-(x - a) / b))

    def linear(x, m, c):
        return m * x + c

    # Select model for curve fitting
    input_data = data['Input'].values
    output_data = data['Output'].values
    if model == 'sigmoid':
        popt, pcov = curve_fit(sigmoid, input_data, output_data, p0=[0, 1, 1])
    else:
        popt, pcov = curve_fit(linear, input_data, output_data)

    return popt, pcov

# Function to plot the results
def plot_results(data, popt, model='sigmoid'):
    """
    Plot synaptic input-output data and the fitted curve.
    
    Args:
        data (DataFrame): Synaptic data containing input and output columns.
        popt (ndarray): Optimal parameters for the fitted curve.
        model (str): The model used for curve fitting ('sigmoid' or 'linear').
    """
    # Define model functions for plotting
    def sigmoid(x, a, b, c):
        return c / (1 + np.exp(-(x - a) / b))

    def linear(x, m, c):
        return m * x + c

    # Plot data points
    plt.scatter(data['Input'], data['Output'], label='Data', color='blue')

    # Generate points for fitted curve
    x_fit = np.linspace(min(data['Input']), max(data['Input']), 100)
    y_fit = sigmoid(x_fit, *popt) if model == 'sigmoid' else linear(x_fit, *popt)

    # Plot fitted curve
    plt.plot(x_fit, y_fit, label=f'Fitted {model.capitalize()} Curve', color='red')
    plt.xlabel('Synaptic Input')
    plt.ylabel('Synaptic Output')
    plt.title('Synaptic Input-Output Analysis')
    plt.legend()
    plt.show()

# Main function to run the analysis
if __name__ == "__main__":
    try:
        # Load the data
        data = load_data('../../data/synaptic_io_data.csv')  # Ensure the data file exists in the 'data/' directory
        
        # Perform the analysis
        popt, pcov = analyze_synaptic_io(data, model='sigmoid')

        # Plot the results
        plot_results(data, popt, model='sigmoid')

    except Exception as e:
        print(f"An error occurred: {e}")